### ISSUE TITLE
Gather feature!

### ISSUE THREAD

**Issue by TangJiakai** [OpenRLHF/OpenRLHF#504]
Currently all_gather only supports tensor objects.
I need a api for gather other non-tensor objects

**Comment by TangJiakai** [comment]
Like this
```
def gather_for_metrics(self, input_data, use_gather_object=False):
        """
        Gathers `input_data` and potentially drops duplicates in the last batch if on a distributed system. Should be
        used for gathering the inputs and targets for metric calculation.

        Args:
            input (`torch.Tensor`, `object`, a nested tuple/list/dictionary of `torch.Tensor`, or a nested tuple/list/dictionary of `object`):
                The tensors or objects for calculating metrics across all processes
            use_gather_object(`bool`):
                Whether to forcibly use gather_object instead of gather (which is already done if all objects passed do
                not contain tensors). This flag can be useful for gathering tensors with different sizes that we don't
                want to pad and concatenate along the first dimension. Using it with GPU tensors is not well supported
                and inefficient as it incurs GPU -> CPU transfer since tensors would be pickled.

        Example:

        ```python
        >>> # Assuming two processes, with a batch size of 5 on a dataset with 9 samples
        >>> import torch
        >>> from accelerate import Accelerator

        >>> accelerator = Accelerator()
        >>> dataloader = torch.utils.data.DataLoader(range(9), batch_size=5)
        >>> dataloader = accelerator.prepare(dataloader)
        >>> batch = next(iter(dataloader))
        >>> gathered_items = accelerator.gather_for_metrics(batch)
        >>> len(gathered_items)
        9
        ```
        """
```

**Comment by tsaoyu** [comment]
What are you trying to do? 
You can always convert your value to a tensor and using collective comm. as long as they are in the same comm. group.

**Comment by TangJiakai** [comment]
I want to, in a distributed situation, load data into the dataloader, perform inference, and then obtain all generated string data, and transfer it to rank-0 to convert to CPU for metric calculation. However, it currently seems not simple to achieve.

**Comment by TangJiakai** [comment]
I am dealing directly with strings, so I cannot use strategy.all_gather.
